tests/src/PatternAssertion/PaginationPatternAssert.php

```diff
@@ -0,0 +1,115 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\Tests\oe_bootstrap_theme\PatternAssertion;
+
+use Symfony\Component\DomCrawler\Crawler;
+
+/**
+ * Assertions for the pagination pattern.
+ */
+class PaginationPatternAssert extends BasePatternAssert {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function getAssertions(string $variant): array {
+    return [
+      'links' => [
+        [$this, 'assertLinks'],
+      ],
+      'alignment' => [
+        [$this, 'assertAlignment'],
+      ],
+      'size' => [
+        [$this, 'assertSize'],
+      ],
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function assertBaseElements(string $html, string $variant): void {
+    $this->assertCounts([
+      'nav' => 1,
+      'ul' => 1,
+    ], new Crawler($html));
+  }
+
+  /**
+   * Asserts the pager links.
+   *
+   * @param array[] $expected
+   *   Expected pager links.
+   *   Each link is an array, with the following keys:
+   *     - 'url': The link destination.
+   *     - 'label': The label, if exists.
+   *     - 'icon': The name of the icon.
+   *     - 'active': Whether the item has an 'active' class.
+   * @param \Symfony\Component\DomCrawler\Crawler $crawler
+   *   The crawler.
+   */
+  protected function assertLinks(array $expected, Crawler $crawler): void {
+    $expected_bcl_icon_url = '/' . \Drupal::service('extension.list.theme')->getPath('oe_bootstrap_theme') . '/assets/icons/bcl-default-icons.svg';
+
+    $actual_links_data = array_map(
+      function (\DOMElement $actual_li_element) use ($expected_bcl_icon_url) {
+        $li = new Crawler($actual_li_element);
+        $link = $li->filter('a');
+        $actual_link_data = [];
+        $actual_link_data['url'] = $link->attr('href');
+        $use = $li->filter('a > svg > use');
+        if ($use->count()) {
+          // This is an icon.
+          $icon_url = $use->attr('xlink:href');
+          [$actual_bcl_icon_url, $actual_link_data['icon']] = explode('#', $icon_url, 2) + [NULL];
+          $this->assertSame($expected_bcl_icon_url, $actual_bcl_icon_url);
+        }
+        else {
+          $actual_link_data['label'] = $link->html();
+        }
+        if ($li->filter('li.active')->count()) {
+          $actual_link_data['active'] = TRUE;
+        }
+        return $actual_link_data;
+      },
+      iterator_to_array($crawler->filter('nav > ul > li'), FALSE),
+    );
+
+    $this->assertSame($expected, $actual_links_data);
+
+    $expected_count = count($expected);
+    $this->assertCounts([
+      'li' => $expected_count,
+      'a' => $expected_count,
+      'nav > ul > li.page-item > a.page-link:first-child' => $expected_count,
```

Comment:
My strategy with count asserts is like this:
- First count with the broadest selector possible, to catch all `<a>` or `<li>` tags, including those in unexpected places or with missing classes.
- Then count again with strictest selectors for position/structure and classes/attributes.

If both counts are the same, we know that all existing `<a>` or `<li>` tags are in the expected position and have the expected attributes.

The `:first-child` makes sure that every `<li>` has exactly one `<a>`. E.g. it would fail if we have one empty `<li></li>` and then one `<li><a></a><a></a></li>` with more than one link tag.

"My bag contains exactly 3 items of clothing and 15 buttons, _and_ my bag contains 3 yellow shirts size M each having 5 buttons." -> no loose leftover buttons, no other items of clothing.